608 dx fix eslint config

christophermiles opened the pull request:
This PR adds the [official Nuxt ESLint module](https://eslint.nuxt.com/packages/module) in place of the previous manual ESLint configuration. This means we will now be using the more modern flat ESLint config.

#### Removal of Prettier

Prettier has been removed as part of this change. We were only using the single-quotes rule from Prettier, and a plugin for Tailwind formatting. 

The single quote rule can also be achieved using [https://eslint.style](https://eslint.style), which is built into the Nuxt ESLint module (see [https://eslint.nuxt.com/packages/module#eslint-stylistic](https://eslint.nuxt.com/packages/module#eslint-stylistic)).

The Tailwind plugin we were using was also available as `eslint-plugin-tailwindcss` so I have incorporated that into the ESLint config to achieve the same result.

#### Removal of SASS files

I have also transposed the SASS CSS to the main `tailwind.css` file.

#### Additions from recent PRs

The addition of the recent `useFilterState` PR that went into staging is not yet present in the target `feature/v2-api` branch so the changes from the `useFilterState` branch will appear in the change list of this PR

#### Formatting changes

The bulk of the file changes are a result of running `lintfix` over the files with the new ESLint config. Mostly these are the removal of semi-colons.

#### Other changes

- The linter suggested using `never` in place of `any` in some TypeScript definitions
- The Tailwind linter recommended combining some classes like `w-6 h-6` into `size-6`, and made other suggestions and fixes

calumbell committed:
Documents V2 (#557)

* SourceModal now pulls srcs from v2

* added transformer to convert V2 API source keys to V1 keys

* updated API result display cmpnts to correctly render links

* fixed bugs in v1/v2 source transformer

* searching now working with v2 documents

* added ruleset to SourcesModal

* refactored useAllSpells composable

* fields now passed to useDocuments composable in SourceModal cmpnt

eepMoody committed:
Enhance document fetching with additional metadata

eepMoody committed:
rework API to use new endpoints, update spell list & view pages

eepMoody committed:
fix spell rendering for material components

calumbell committed:
Feature: V2 Monsters (#562)

* /monster now pulls from V2

* monsters/[id] now pulls from v2

* fixed source selection of /monsters route

* rmv'd trailing 0s from CR on /monsters

* monster filter works with V2 (mostly)

* re-added SourceTags to ApiResultRow

calumbell committed:
fixed linted errors on monsters/[id] (#563)

calumbell committed:
Feature: V2 Magic Items (#564)

* /magic-items top level page now pulls from v2

* api composable now points to v2 items endpoint

* added requires_attunement column to /magic-items

* magic-items/[id] works with v2/items endpoint

* /magic-items list requires attunement filter working

* added comments to magic item filters

calumbell committed:
Feature: V2 Breadcrumb (#565)

* refactored breadcrumbing into useBreadcrumb composable

* BreadcrumbLinks now handles V2 keys more gracefully

* updated page title generation w/ breadcrumbs

calumbell committed:
bugfix: /monsters filtering and sorting (#570)

calumbell committed:
API V2 - Backgrounds (#571)

* switched to backgrounds v2 endpoint

* sorted background features to improve layout

calumbell committed:
All sources navigable on initial page load (#572)

calumbell committed:
V2 Classes (#574)

* migrated /classes route to API V2

* /classes/[className] now pulls from API V2

* added subclass links to classes

* updated class links in navbar

* class features now ordered correctly on /classes/[className] route

* fixed bug on pages/classes/[className] route

* updated useFindOne composable to accept query parameters

* subclasses now appear as subroute of their base-class

* removed unneeded fetching from /classes/[className]/[subclass]

* removed console log from composables/api

* updated navbar to highlight base class when subclass of it selected

* added comments explaining API doesn't currently return spell slots

* hide source tags in breadcrumb on small screen widths

calumbell committed:
Refactor: move data fetching from `ApiResultsTable` component to page (#575)

* moved useFindPaginated to own file

* refactored ApiResultsTable to move fetching logic onto page - tested on /monsters route

* refactored ApiTableNav - more DRY

* fixed bugs with ApiTableNav

* refactored /magic-items

* refactored /spells

* refactored /classes

* refactored /backgrounds page

* added comments js on pages

* added comments to new composables

calumbell committed:
Feature: V2 Races (#576)

* /races now pulls from API V2

* refactored /races/[id] route

* added markdown rendered to /races/[id]

calumbell committed:
Feature: V2 Feats (#577)

calumbell committed:
Feature: V2 Conditions (#578)

* updated /conditions and /conditions/[id] to Open5e API v2

* removed relict 'endpoint' prop from ApiResultsRow/Table cmpnts

* fixed bug in conditions 'fields' query param

calumbell committed:
Feature: API V2 - Merge & Rebase (#579)

* 539 Bug: trailing comma in monsters without alignment (#543)

* remove trailing comma

* remove extra space

* refactor

* updated eslint rules to permit one-line conditionals (#573)

---------

Co-authored-by: Ean Moody <moody.designer@gmail.com>

calumbell committed:
refactored single-item page units tests to work with API V2 data

calumbell committed:
API V2: Test (#580)

* updated tests for top-level routes to API V2

* added missing background tests

calumbell committed:
V2 Feature: Search (#582)

* updated /search to use v2 keys - subclass links now work correctly

* improved subtitles on /search results

* improved comments on SearchResult cmpnt

* add early rtrns to work more gracefully w/ current API staging branch

calumbell committed:
API V2: Rulesets (#586)

* implemented UI for ruleset selector

* ruleset now filters sources in SourceModal

* ruleset state now persists

* fixed bug with loading undefined rulesets

* simplified onRulesetChanged()

* updated all/none controls to work with ruleset

* added comments to SourcesModal

calumbell committed:
V2 - Magic Item Filters (#587)

calumbell committed:
Feature: V2 Class Spells Filter (#589)

* /spells displays classes instead of components

* renamed composables/filters.ts -> composables/useFilterState.ts

* /spells table now filterable

* updated UI for spell filters

* removed spells by class pages from navbar

* added spell classes to spells/[id]

calumbell committed:
updated references to ruleset to gamesystem (#591)

calumbell committed:
API V2: Rules (#593)

* removed relict pages that pulled from /sections endpoint

* tidied up default layout

* added /rules page to display all rulesets from API

* added /rules/[id] route for displaying single rulesets

calumbell committed:
API V2: Upgrade Filters (#595)

* /spells now uses debounced filter

* added ApiTableFilter cmpnt to capture filtering logic

* refactored /magic-items to use new filters

* restyled /magic-items to better match /spells UI

* refactored /spells using ApiTableFilter cmpnt

* refactored /monsters using ApiTableFilter cmpnt

* removed relict filtering components

* updated remaining pages to match new UI, added search where supported

calumbell committed:
API V2: Equpiment (#597)

* added /equipment page

* added /equipment/[id] route

* improve html semantics on equipment/[id]

* fixed search links to equipment/[id]

* added weapon detail view to /equipment/[id]

* added armor detail view to /equipment/[id]

* updated styled and fixed corner cases for /equipment/[id] page

calumbell committed:
API V2: Redirects (#598)

* updated redirect middleware to better handle likely v1/v2 collisions

* added redirects to data fetching composables when a dynamic route 404 is rtnd

calumbell committed:
API call on /spells now requests correct fields from API (added missing join) (#599)

calumbell committed:
API V2: bugfix, query params for /backgrounds route (#600)

calumbell committed:
Update fetching of nested parameters (#603)

* fixed overfetching on layout (getting document count)

* updated query params to reduce overfetching on top-level pages

* ApiResultsRow cmpnt now uses 'key' field for links (no need to extract key from url)

* added reduced overfetching in call to /documents in SourcesModal component

* bugfix: added query params to useQuery dependencies in useDocuments()

christophermiles committed:
Remove duplicate div declaration

christophermiles committed:
Typo

calumbell committed:
API V2: Merge from `Staging` (#605)

* 539 Bug: trailing comma in monsters without alignment (#543)

* remove trailing comma

* remove extra space

* refactor

* updated eslint rules to permit one-line conditionals (#573)

* Build(deps): Bump serve-static from 1.15.0 to 1.16.2 (#583)

Bumps [serve-static](https://github.com/expressjs/serve-static) from 1.15.0 to 1.16.2.
- [Release notes](https://github.com/expressjs/serve-static/releases)
- [Changelog](https://github.com/expressjs/serve-static/blob/v1.16.2/HISTORY.md)
- [Commits](https://github.com/expressjs/serve-static/compare/v1.15.0...v1.16.2)

---
updated-dependencies:
- dependency-name: serve-static
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Build(deps): Bump vite from 5.4.1 to 5.4.6 (#584)

Bumps [vite](https://github.com/vitejs/vite/tree/HEAD/packages/vite) from 5.4.1 to 5.4.6.
- [Release notes](https://github.com/vitejs/vite/releases)
- [Changelog](https://github.com/vitejs/vite/blob/v5.4.6/packages/vite/CHANGELOG.md)
- [Commits](https://github.com/vitejs/vite/commits/v5.4.6/packages/vite)

---
updated-dependencies:
- dependency-name: vite
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Build(deps): Bump rollup from 4.21.0 to 4.24.0 (#594)

Bumps [rollup](https://github.com/rollup/rollup) from 4.21.0 to 4.24.0.
- [Release notes](https://github.com/rollup/rollup/releases)
- [Changelog](https://github.com/rollup/rollup/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rollup/rollup/compare/v4.21.0...v4.24.0)

---
updated-dependencies:
- dependency-name: rollup
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Ean Moody <moody.designer@gmail.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

calumbell committed:
Merge branch 'staging' into feature/v2-api

christophermiles committed:
Sync current filter state to local storage

christophermiles committed:
Persist checkbox status

christophermiles committed:
Rename

christophermiles committed:
Merge branch 'staging' into 529-ux-retain-filter-settings-when-navigating-back-to-monsters-page

christophermiles committed:
Fixed 500 error caused by local storage being initially undefined

christophermiles committed:
Removed unused function

christophermiles committed:
Reflect persisted search value in search bar

Also added TODO note

christophermiles committed:
Merge branch 'feature/v2-api' into 529-ux-retain-filter-settings-when-navigating-back-to-monsters-page

christophermiles committed:
Fixed filtering on other pages and simplified API for filtering

christophermiles committed:
Aligned clear button presentation with pagination buttons

christophermiles committed:
Persist filter across session but not on refresh

christophermiles committed:
Local storage key no longer required for filter state

christophermiles committed:
Formatting

christophermiles committed:
Track multiple filter fields states from singleton

Previously, filter state was wiped from one filter when initialising a new one

christophermiles committed:
Fix

christophermiles committed:
Removed unused type declaration

christophermiles committed:
Make filters more type safe

christophermiles committed:
Removed unused duplicate

calumbell committed:
API V2: Update query params for `/monsters/[id]` route (#609)

* fixed languages & environments in mosnter statblocks. reduced overfetching

* fixed environment overflow - sections now respect width of container class

christophermiles committed:
Fixed incorrect access

christophermiles committed:
Added reusable button for table controls

christophermiles committed:
Merge branch 'feature/v2-api' into 529-ux-retain-filter-settings-when-navigating-back-to-monsters-page

christophermiles committed:
First pass at config updates and linter fixes

christophermiles committed:
Merge branch '529-ux-retain-filter-settings-when-navigating-back-to-monsters-page' into 608-dx-fix-eslint-config

christophermiles committed:
Some progress

christophermiles committed:
Linting fixes

christophermiles committed:
529 ux retain filter settings when navigating back to monsters page (#607)

Search, select and checkbox filters for API results pages are now persisted during the session. This applies to the following pages:

* Classes
* Equipment
* Magic Items
* Monsters
* Rules
* Spells

Additionally, a 'Clear filter' button will become available while a page has an active filter.

christophermiles committed:
Formatting

christophermiles committed:
Fix scripts

christophermiles committed:
Typing props

christophermiles committed:
Fixed classname

christophermiles committed:
Typing filter state

christophermiles committed:
Trying to get type safety working on filter

christophermiles committed:
Export type

christophermiles committed:
Fixing generic usage

christophermiles committed:
Transposed SASS styling to Tailwind and CSS vars

christophermiles committed:
Whoops

christophermiles committed:
Merge branch 'feature/v2-api' into 608-dx-fix-eslint-config

christophermiles committed:
Lint fix

christophermiles commented on .husky/pre-commit:
```diff
@@ -1,9 +1,6 @@
 
-# Run pretty-quick to format staged files
-npx pretty-quick --staged
-
```
I removed this because we're no longer using Prettier, but wasn't sure how best to replace this

christophermiles commented on bin/generate.js:
- Removed unused imports
- Lint fixes

christophermiles commented on composables/api.ts:
`any` has been changed to `never` at the suggestion of the linter

christophermiles commented on styles/tailwind.css:
I've moved the CSS declarations from the SASS files here

I think we could use the Tailwind prose plugin in place of a lot of the rules for setting spacing between headings.

I think there is probably more we could do to consolidate the CSS and leverage Tailwind, but that is probably outside the scope of this PR.

calumbell commented:
@christophermiles this looks great, nice work! It is also pretty big. It will take a bit of time to review, but I am on it

christophermiles commented:
@calumbell No worries, and no rush. I realise nobody actually asked for this but I thought it might be nice to update the ESLint and remove the SASS dependency

calumbell reviewed (changes requested):
There is a lot in here. First up, colapsing the SASS into Tailwind classes looks like it has gone well. And it looks like the TS configuration is working properly now, lots of the annoying non-error TypeScript errors have vanished. Nice works.

There are still a few problems with the linter that I noticed:

1) The semi-colons are gone, could we put them back in? 

2) Single and double quotes appear to be handled differently in different context. IE. in the `ClassTable` component, single-quotes in the `<template>` tag were changed to double quotes, but not the in `<script>` tag. This happens is several (all?) other components and composables. My instinct is that this should be consistent across the codebase. Can you update the ESLint rule to enforce one or the other across the whole site?

calumbell commented:
@christophermiles Oh another thing! The branch you are pushing to here, `feature/v2-api`, has been merged to staging. Probably easiest if push this PR to `staging` also.

Thanks for your work!

christophermiles committed:
Restored semicolons

christophermiles commented:
@calumbell I've fixed the semi-colons, apologies for the oversight.

With regard to the quotes, the Nuxt ESLint module uses the Vue recommended settings by default, which includes the use of double quotes in templates as the default: [https://eslint.vuejs.org/rules/html-quotes.html](https://eslint.vuejs.org/rules/html-quotes.html)

I have brought across any overrides we previously had in our ESLint config to the new config, and there doesn't seem to be anything explicitly in there about setting quotes to single in templates.

I'm not sure if I've ever seen a Vue project use single quotes in template blocks before, as there is often a need to do some string interpolation within attribute declaration, which (I assume?) would be made more difficult if using single quotes. 

But happy to override the default if there is some reason for using single quotes in the template code that I'm not aware of?

christophermiles commented:
@calumbell Unless with the quotes you meant stuff like this:

```
<<<<<<< HEAD
        <span>{{ col.value(data) ? "√" : "-" }}</span>
=======
        <span>{{ col.value(data) ? '√' : '-' }}</span>
>>>>>>> staging
```

In which case, yeah, I think it does make more sense for these to be single quote since these are JavaScript expressions. I thought you might have meant single quotes for the attributes.

I'll have a look to see if there's a setting for single quotes in JS expressions in templates.

christophermiles committed:
Merge branch 'staging' into 608-dx-fix-eslint-config

christophermiles committed:
Use 'one true brace style' setting

christophermiles commented:
@calumbell Apologies for the delay, have made those changes now. 

I also made a change since the last PR to use the ['one true brace style' setting for the ESLint stylistic plugin](https://eslint.style/rules/js/brace-style#_1tbs), which was defaulting to ['stroustrup'](https://eslint.style/rules/js/brace-style#stroustrup)

It seemed like 'one true brace style' was closer to the convention being used (and is certainly closer to what I would normally use!)

christophermiles commented:
@calumbell @eepMoody I see one of the automated checks is failing though — I had a log at the logs and it seems to be something to do with some Python code?

calumbell reviewed (approved):
Hey @christophermiles, excuse the delay in getting this reviewed. I have been busy and this PR is a monster. Thanks for your contribution, but if we could keep future PRs a bit smaller and focused on single issues we'll be able to get these reviewed much faster.

The DX changes appear to be good! Poking around on local I encounter way fewer weird non-errors.

Many of the Tailwind changes are positive, but I have a few concerns (i'll slap these into their own issue, they are way beyond the scope of this PR). TLDR if we are using Tailwind, we should use be Tailwind classes and not the old classes copy-pasted from the SASS file to the Tailwind config. It buries them a bit to deeply and makes them (in my opinion) more difficult to maintain.

Thanks again for your contribution! Enough time has passed that we have some merge conflicts in our dependencies. Hopefully will only take a merge and a `npm i` to resolve

calumbell commented:
> @calumbell @eepMoody I see one of the automated checks is failing though — I had a log at the logs and it seems to be something to do with some Python code?

Yeah, the automated checks have been giving me problems as well. I also haven't gotten any useful feedback from them. I think we should get rid of them, what do you think @eepMoody ?

calumbell commented:
(Closing and reopening to re-run tests)

calumbell committed:
resolved merge conflicts with 'staging' branch

eepMoody commented:
> > @calumbell I think we should get rid of them, what do you think?

Yeah go for it! It was an experiment but it clearly isn't producing anything useful.

calumbell commented:
@christophermiles Apologies for how long this has taken to merge. I have submitted a PR to remove the AI code reviewer (#678), which is not working and blocking the merge. Once this is approved we should be able to finally merge these changes to `staging`.

eepMoody commented:
@calumbell do you want me to just override the protections and get this merged in? I hadn't ever intended that AI code review thing to be a blocker 🙏 sry

calumbell committed:
merged from staging, fixed conflicts

calumbell committed:
fixed bugs that occured due to merge

calumbell committed:
regenerated package-lock.json to fix opt. dep. bug
